README.md
Password Strength Tester
========================

Password Strength is a library that calculates the relative strength of a password.  This is accomplished by using several techniques.  Primarily this relies on letter trigraphs, which check each set of 3 characters in a given password.  More information on the trigraph calculations [is available](data/README.md).  This also calculates the entropy bits based on Claude Shannon's technique on determining the number of bits required to represent a set of characters and multiplying it by the length of the password.  There is also a check to see if a password is contained in a list of common passwords.

There is a wonderful [demo page](http://tests-always-included.github.io/password-strength) where you can test out the technology yourself.

[![npm version][npm-badge]][npm-link]
[![Build Status][travis-badge]][travis-link]
[![Dependencies][dependencies-badge]][dependencies-link]
[![Dev Dependencies][devdependencies-badge]][devdependencies-link]


Usage
-----

The Password Strength library is wrapped with [fid-umd], and is usable in a variety of systems.  When you use the `.check()` method for the password "abcd1234", the results may look like what you see below.  Comments are added to explain the fields.

    {
        charsetSize: 36,  // Explained better below
        commonPassword: true,  // If true, don't use this password!
        nistEntropyBits: 18, // NIST 800-63, probably not a good measurement
        passwordLength: 8,  // Same as string.length
        shannonEntropyBits: 24,  // Claude Shannon's method
        strengthCode: 'WEAK',  // Our ranking of the password's strength
        trigraphEntropyBits: 39.71755017780513,  // Based on trigraphs
        charsets: {
            number: true,  // Contains 0-9
            lower: true,  // Contains a-z
            upper: false,  // Contains A-Z
            punctuation: false,  // Contains common sentence punctuation
            symbol: false,  // Contains mathematical symbols
            other: ''  // Unicode and uncaught characters
        }
    }

The charset size is the sum of the lengths of the different charsets that the password uses.  The higher this number, the harder it is to brute force attack.  That's precisely why password policies often say "must contain one lowercase letter, one uppercase letter, one number and a symbol".

The strength code is based on the trigraph entropy bits when they are available and will fall back to the Shannon entropy bits.  It's one of five values:  `VERY_WEAK`, `WEAK`, `REASONABLE`, `STRONG`, and `VERY_STRONG`. The NIST scores are based on NIST Special Publication 800-63, but this is considered to not provide an accurate measurement of entropy and is included for completeness and comparison to the other algorithms.

Trigraph entropy bits is discussed more [here](data/README.md).

The charsets is mostly an object that has boolean values, except the `other` property.  That one is a catch-all string of the letters that were not caught and tallied into another one of the charset lists.  The list of characters in `other` is deduplicated.


### Browser

Include `lib/password-strength.js` in your project.

    <script src="path/to/password-strength.js" />

Next you will want to instantiate the modules and make some AJAX calls to fetch additional data and make the password strength tester more accurate and informative.  This bit of code uses jQuery, but similar code can be written for any framework.

    $(function () {
        // Create the instance that can be used immediately for strength tests
        window.passwordStrength = new PasswordStrength();

        // Add additional files that improve the results
        $.getJSON("path/to/data/common-passwords.json", function (data) {
            window.passwordStrength.addCommonPasswords(data);
        });
        $.getJSON("path/to/data/trigraphs.json", function (data) {
            window.passwordStrength.addTrigraphMap(data);
        });
    });

Later, to calculate the strength of a password you would use something like this:

    if (window.passwordStrength) {
        strength = window.passwordStrength.check("abcd1234");

        if (strength.strengthCode.indexOf('WEAK') >= 0) {
            alert("Your password is too weak.");
        }
    }


### Node.js

First run `npm install --save tai-password-strength` and then your code would look a bit like this:

    var taiPasswordStrength = require("tai-password-strength")
    var strengthTester = new taiPasswordStrength.PasswordStrength();
    var results = strengthTester.check("abcd1234");

    // Add in extra files for additional checks and better results
    strengthTester.addCommonPasswords(taiPasswordStrength.commonPasswords);
    strengthTester.addTrigraphMap(taiPasswordStrength.trigraphs);
    var betterResults = strengthTester.check("abcd1234");

    if (betterResults.strengthCode.indexOf('WEAK') >= 0) {
        throw new Error("Your password is too weak");
    }


[dependencies-badge]: https://img.shields.io/david/tests-always-included/password-strength.svg
[dependencies-link]: https://david-dm.org/tests-always-included/password-strength
[devdependencies-badge]: https://img.shields.io/david/dev/tests-always-included/password-strength.svg
[devdependencies-link]: https://david-dm.org/tests-always-included/password-strength#info=devDependencies
[fid-umd]: https://github.com/fidian/fid-umd
[npm-badge]: https://img.shields.io/npm/v/tai-password-strength.svg
[npm-link]: https://npmjs.org/package/tai-password-strength
[travis-badge]: https://img.shields.io/travis/tests-always-included/password-strength/master.svg
[travis-link]: http://travis-ci.org/tests-always-included/password-strength


CHANGELOG.md
CHANGELOG
=========


1.1.2 - 2021-03-02
------------------

* Added TypeScript, per request.


1.1.1 - 2017-08-24
------------------

* Fixed example in `README.md`.
* Switched from grunt, jasmine, and jslint to running jasmine with eslint.
* No functional changes.


1.1.0 - 2017-06-27
------------------

* Added NIST algorithm.
* Fixed NPM package name in `README.md`.
* Added more common passwords.
* Updated badge links in `README.md`.
* Additional documentation about entropy measurements.


1.0.4 - 2016-05-16
------------------

* Fixes broken links in `package.json`.
* Added links to live site in `README.md`.


1.0.3 - 2016-04-05
------------------

* Added extra tests.


1.0.2 - 2016-04-05
------------------

* Fixing the XKCD magic password.


1.0.1 - 2016-03-25
------------------

* Added comma to the list of punctuation characters.


1.0.0 - 2016-03-23
------------------

* Initial release.


spec/lib/password-strength.spec.js
"use strict";

var PasswordStrength, passwordStrength;


beforeEach(function () {
    PasswordStrength = require("../../").PasswordStrength;
    passwordStrength = new PasswordStrength();
});

describe("addCommonPasswords()", function () {
    it("adds data as JSON then passes nothing to reset", function () {
        passwordStrength.addCommonPasswords([
            "abc123",
            "123456"
        ]);

        expect(passwordStrength.commonPasswords).toEqual([
            "abc123",
            "123456"
        ]);

        passwordStrength.addCommonPasswords();

        expect(passwordStrength.commonPasswords).toEqual([]);
    });

    it("adds data as a string", function () {
        passwordStrength.addCommonPasswords("abc123\n123456");

        expect(passwordStrength.commonPasswords).toEqual([
            "abc123",
            "123456"
        ]);
    });

    it("tries to add another type of format", function () {
        // Need to wrap the method in a function call so we can catch the error
        expect(function () {
            passwordStrength.addCommonPasswords({
                abc123: true,
                123456: false
            });
        }).toThrow();
    });
});


describe("addTrigraphMap()", function () {
    it("passes a proper trigraph set and then a blank one", function () {
        passwordStrength.addTrigraphMap({
            _qu: 9468,
            quo: 267,
            uot: 2820,
            te_: 1678
        });

        expect(passwordStrength.trigraph).toEqual({
            _qu: 9468,
            quo: 267,
            uot: 2820,
            te_: 1678
        });

        passwordStrength.addTrigraphMap();
        expect(passwordStrength.trigraph).toEqual(null);
    });

    it("tries to add an array", function () {
        // Need to wrap the method in a function call so we can catch the error
        expect(function () {
            passwordStrength.addTrigraphMap(["_qu", "quo"]);
        }).toThrow();
    });
});

describe("charsetSize()", function () {
    it("determines size with lowercase and uppercase", function () {
        expect(passwordStrength.charsetSize({
            lower: true,
            upper: true,
            number: false,
            punctuation: false,
            symbol: false
        })).toBe(52);
    });

    it("determines size with numeric, punctuation and symbols", function () {
        expect(passwordStrength.charsetSize({
            lower: false,
            upper: false,
            number: true,
            punctuation: true,
            symbol: true
        })).toBe(42);
    });
});

describe("charsetGroups()", function () {
    it("talllies uppercase and numbers", function () {
        expect(passwordStrength.charsetGroups("00DEADBEEF00")).toEqual({
            number: true,
            lower: false,
            upper: true,
            punctuation: false,
            symbol: false,
            other: ""
        });
    });
    it("detects lowercase, punctuation and symbols", function () {
        expect(passwordStrength.charsetGroups("p@sswo.rd")).toEqual({
            number: false,
            lower: true,
            upper: false,
            punctuation: true,
            symbol: true,
            other: ""
        });
    });
    it("supports Unicode characters", function () {
        var airplane, star;

        airplane = String.fromCharCode(9992);
        star = String.fromCharCode(9728);
        expect(passwordStrength.charsetGroups(star + "fdfdfd" + airplane)).toEqual({
            number: false,
            lower: true,
            upper: false,
            punctuation: false,
            symbol: false,
            other: star + airplane
        });
    });
    it("deduplicates Unicode characters", function () {
        var airplane, star;

        airplane = String.fromCharCode(9992);
        star = String.fromCharCode(9728);
        expect(passwordStrength.charsetGroups(star + star + airplane + star + airplane)).toEqual({
            number: false,
            lower: false,
            upper: false,
            punctuation: false,
            symbol: false,
            other: star + airplane
        });
    });
});

describe("check()", function () {
    beforeEach(function () {
        passwordStrength.addCommonPasswords([
            "abc123",
            "abcd1234",
            "quote"
        ]).addTrigraphMap({
            _qu: 9468,
            quo: 267,
            uot: 2820,
            te_: 1678
        });
    });

    it("returns all calculated status properties for quote", function () {
        expect(passwordStrength.check("quote")).toEqual({
            charsetSize: 26,
            commonPassword: true,
            nistEntropyBits: 12,
            passwordLength: 5,
            shannonEntropyBits: 11.60964047443681,
            strengthCode: "VERY_WEAK",
            trigraphEntropyBits: 18.487783049487163,
            charsets: {
                number: false,
                lower: true,
                upper: false,
                punctuation: false,
                symbol: false,
                other: ""
            }
        });
    });

    it("returns all calculated status properties for Quot", function () {
        expect(passwordStrength.check("Quot")).toEqual({
            charsetSize: 52,
            commonPassword: false,
            nistEntropyBits: 10,
            passwordLength: 4,
            shannonEntropyBits: 8,
            strengthCode: "VERY_WEAK",
            trigraphEntropyBits: 18.05234113785076,
            charsets: {
                number: false,
                lower: true,
                upper: true,
                punctuation: false,
                symbol: false,
                other: ""
            }
        });
    });
});

describe("checkCommonPasswords()", function () {
    describe("with some common passwords", function () {
        beforeEach(function () {
            passwordStrength.addCommonPasswords([
                "abc123",
                "!helpme"
            ]);
        });
        it("detects a password in the list", function () {
            expect(passwordStrength.checkCommonPasswords("abc123")).toBe(true);
        });
        it("with a password not on the list", function () {
            expect(passwordStrength.checkCommonPasswords("abc1234")).toBe(false);
        });
    });
    describe("without any common passwords", function () {
        it("always returns null", function () {
            expect(passwordStrength.checkCommonPasswords("abc1234")).toBe(null);
        });
    });
});

describe("checkTrigraph()", function () {
    describe("with a trigraph map", function () {
        beforeEach(function () {
            passwordStrength.addTrigraphMap({
                __q: 22,
                _qu: 9468,
                quo: 267,
                uot: 2820,
                ote: 1095,
                te_: 1678
            });
        });
        it("calculates the strength of a password that's fully defined", function () {
            expect(passwordStrength.checkTrigraph("quote", 26)).toBeCloseTo(18.320470565926694);
        });
        it("can work when portions are not in the frequency table", function () {
            expect(passwordStrength.checkTrigraph("quahe", 53)).toBeCloseTo(24.407172328767736);
        });
        it("does not require any matches in the frequency table", function () {
            // Result should be equal to the length * chaset size.
            expect(passwordStrength.checkTrigraph("blAh", 71)).toBeCloseTo(24.59898847801873);
        });
    });
    describe("without a trigraph map", function () {
        it("returns null", function () {
            expect(passwordStrength.checkTrigraph("quote", 26)).toBe(null);
        });
    });
});


describe("determineStrength()", function () {
    it("returns VERY_WEAK from trigraphEntropyBits", function () {
        expect(passwordStrength.determineStrength({
            trigraphEntropyBits: 31
        })).toBe("VERY_WEAK");
    });

    it("returns WEAK from trigraphEntropyBits", function () {
        expect(passwordStrength.determineStrength({
            trigraphEntropyBits: 36
        })).toBe("WEAK");
    });

    it("returns REASONABLE from trigraphEntropyBits", function () {
        expect(passwordStrength.determineStrength({
            trigraphEntropyBits: 53
        })).toBe("REASONABLE");
    });

    it("returns STRONG from shannonEntropyBits", function () {
        expect(passwordStrength.determineStrength({
            trigraphEntropyBits: null,
            shannonEntropyBits: 72
        })).toBe("STRONG");
    });

    it("returns VERY_STRONG from shannonEntropyBits", function () {
        expect(passwordStrength.determineStrength({
            trigraphEntropyBits: null,
            shannonEntropyBits: 95
        })).toBe("VERY_STRONG");
    });
});


describe("nistScore()", function () {
    it("scores an empty string", function () {
        expect(passwordStrength.nistScore("")).toBe(0);
    });
    it("scores a single character", function () {
        expect(passwordStrength.nistScore(".")).toBe(4);
    });
    it("scores a second character", function () {
        expect(passwordStrength.nistScore(".!")).toBe(6);
    });
    it("applies a bonus", function () {
        expect(passwordStrength.nistScore(".A")).toBe(12);
    });
    it("scores a ninth character", function () {
        expect(passwordStrength.nistScore("123456789")).toBe(19.5);
    });
    it("scores a long passphrase", function () {
        expect(passwordStrength.nistScore("123456789012345678901")).toBe(37);
    });
});


describe("otherChars()", function () {
    it("returns empty string when all allphanumeric", function () {
        expect(passwordStrength.otherChars("Mary1")).toBe("");
    });
    it("returns Unicode and other unhandled charcters, such as spaces", function () {
        var airplane, star;

        // ✈
        airplane = String.fromCharCode(9992);

        // ☀
        star = String.fromCharCode(9728);

        expect(passwordStrength.otherChars("A " + star + " on an " + airplane)).toBe(star + airplane);
    });
});


describe("shannonScore()", function () {
    it("finds a single character's entropy", function () {
        // 0 bits to represent * length of 1 = 0
        expect(passwordStrength.shannonScore("x")).toBeCloseTo(0);
    });
    it("finds two character's entropy", function () {
        // 1 bit to represent * length of 2 = 2
        expect(passwordStrength.shannonScore("xy")).toBeCloseTo(2);
    });
    it("computes entropy correctly", function () {
        expect(passwordStrength.shannonScore("1223334444")).toBeCloseTo(18.464393446710154);
        expect(passwordStrength.shannonScore("AAAAABBCDE")).toBeCloseTo(19.609640474436812);
    });
});


lib/password-strength.js
/**
 * This is an object oriented approach of rumkin.com/passchk.js
 * Originally written to be self contained, this makes it as an object
 * that can be instantiated.
 */
"use strict";

// fid-umd {"name":"PasswordStrength"}
(function (name, root, factory) {
    /**
     * Determines if something is an object. Use for minification.
     *
     * @param {*} x
     * @return {boolean}
     */
    function isObject(x) {
        return typeof x === "object";
    }

    if ((typeof module)[0] === "o" && isObject(module.exports)) {
        module.exports = factory();
    } else if ((typeof exports)[0] === "o") {
        exports[name] = factory();
    } else if (isObject(root.define) && root.define.amd) {
        root.define(name, [], factory);
    } else if (isObject(root.modulejs)) {
        root.modulejs.define(name, factory);
    } else if (isObject(root.YUI)) {
        root.YUI.add(name, function (Y) {
            Y[name] = factory();
        });
    } else {
        root[name] = factory();
    }

    // eslint-disable-next-line
}("PasswordStrength", this, function () {
    // fid-umd end

    /**
     * Shim the Math object to support log2 if it does not already.
     */
    if (!Math.log2) {
        /**
         * You can emulate any log in any base with log.
         *
         * @param {number} n
         * @return {number}
         */
        Math.log2 = function (n) {
            return Math.log(n) / Math.log(2);
        };
    }


    /**
     * Iterates over an object
     *
     * @param {Object} obj
     * @param {Function} callback(key,value)
     */
    function objectForEach(obj, callback) {
        Object.keys(obj).forEach(function (key) {
            callback(key, obj[key]);
        });
    }


    /**
     * Turns a list of characters into a regular expression that will match
     * whatever you pass in.  There may be trouble generating the pattern,
     * so here we are escaping every character that isn't alphanumeric.
     *
     * @param {string} letters
     * @param {string} flags (defaults to "")
     * @return {RegExp} pattern
     */
    function safeCharMatcher(letters, flags) {
        var pattern;

        pattern = letters.replace(/[\W_]/g, function (x) {
            return "\\" + x;
        });
        pattern = "[" + pattern + "]";

        return new RegExp(pattern, flags || "");
    }


    /**
     * Determine the strength of a password.
     *
     * Usage:
     *
     *     ps = new PasswordStrength();
     *
     *     // Improve checks by adding additional data files - optional
     *     ps.addCommonPasswords(pass);  // data/common-passwords.json
     *     ps.addTrigraphMap(tri);  // data/trigraphs.json
     *
     *     // Calculate password strength
     *     stats = ps.check("my password");
     *
     * @return {PasswordStrength} instance
     */
    function PasswordStrength() {
        if (!(this instanceof PasswordStrength)) {
            return new PasswordStrength();
        }

        this.commonPasswords = null;
        this.trigraph = null;
        this.charsets = {
            number: "0123456789",
            lower: "abcdefghijklmnopqrstuvwxyz",
            upper: "ABCDEFGHIJKLMNOPQRSTUVWXYZ",

            // Things found in sentences
            punctuation: "!'.,:;?&-\" ",

            // Typically math-related
            symbol: "@#$%^*(){}[]><~`_+=|/"
        };


        /**
         * Sets commonPassword property with what a user passes in.
         * Uses it if it's an array, processes it into an array if it's a
         * string and throws an error if it's not a format we can work with.
         *
         * Returns the PasswordCheck object so it can be chained
         *
         * @param {(Array|string)} passwords
         * @return {PasswordStrength} PasswordStrength
         */
        this.addCommonPasswords = function (passwords) {
            if (passwords) {
                if (Array.isArray(passwords)) {
                    this.commonPasswords = passwords;
                } else if (typeof passwords === "string") {
                    this.commonPasswords = passwords.split(/\r\n|\r|\n/);
                } else {
                    throw new Error("Format does not match any expected format.");
                }
            } else {
                this.commonPasswords = [];
            }

            return this;
        };


        /**
         * TrigraphMap is an ojbect representing entries of 3 letter
         * combinations and how often they appear.
         *
         * {
         *     "_ty": 123,
         *     "tyl": 22,
         *     "yle": 19,
         *     "ler": 5,
         *     "er_": 543
         * }
         *
         * For further explanation on the numbers, check out the documentation
         * in the data folder.
         *
         * @typedef {Object} PasswordStrength~trigraphMap
         * @property {number} * Frequencies of that letter combo.
         */

        /**
         * Returns the PasswordStrength object so it can be chained
         *
         * @param {PasswordStrength~trigraphMap} trigraphMap
         * @return {PasswordStrength} PasswordStrength
         */
        this.addTrigraphMap = function (trigraphMap) {
            if (trigraphMap) {
                if (trigraphMap && typeof trigraphMap === "object" && !Array.isArray(trigraphMap)) {
                    this.trigraph = trigraphMap;
                } else {
                    throw new Error("Format does not match any expected format.");
                }
            } else {
                this.trigraph = null;
            }

            return this;
        };


        /**
         * charsetGroups is an object of different group of defined characters
         * used in figuring out how many characters per group as well as
         * what characters fall into each group.  It is based on a
         * PasswordStrength instance's `this.charsets` object.  When a password
         * has a character that matches one of the character sets in the
         * PasswordStrength instance's sets, this sets that key to `true`.
         *
         * The "other" key is all letters that did not match another set.
         *
         * Example for the password "abcd£1234":
         *
         * {
         *     lower: true,
         *     upper: false,
         *     number: true,
         *     punctuation: false,
         *     symbol: false,
         *     other: "£"
         * }
         *
         * @typedef {Object} PasswordStrength~charsetGroups
         * @property {string} other Characters that did not match another set
         * @property {boolean} * All of the rest of the sets just set flags
         */

        /**
         * Gets the character set size of the current password.
         * This breaks up the each character set and tests the password
         * to see which set they belong in.
         *
         * @param {string} passToCheck
         * @return {PasswordStrength~charsetGroups} groups
         */
        this.charsetGroups = function (passToCheck) {
            var groups;

            groups = {};
            objectForEach(this.charsets, function (key, value) {
                groups[key] = safeCharMatcher(value).test(passToCheck);
            });
            groups.other = this.otherChars(passToCheck);

            return groups;
        };


        /**
         * Calculates the size of each flag found for a password,
         * and will get the flags if a password was passed in.
         *
         * @param {PasswordStrength~charsetGroups} groups
         * @return {number} result
         */
        this.charsetSize = function (groups) {
            var size;

            size = 0;

            objectForEach(this.charsets, function (key, value) {
                if (groups[key]) {
                    size += value.length;
                }
            });

            if (typeof groups.other === "string") {
                size += groups.other.length;
            }

            return size;
        };

        /**
         * Statistics is an object with information about the current state of
         * the password.
         *
         * @typedef {Object} PasswordStrength~statistics
         * @property {number} charsetSize size of the characters from each group the characters fall into.
         * @property {boolean} commonPassword
         * @property {number} passwordLength the length of the current password
         * @property {number} shannonEntropyBits
         * @property {string} strength estimates strength score for the password
         * @property {(null|number)} trigraphEntropyBits
        */

        /**
         * Calculates a bunch of statistics about a given password.
         *
         * @param {string} currentPassword
         * @return {PasswordStrength~statistics}
         */
        this.check = function (currentPassword) {
            var result;

            result = {
                charsetSize: 0,
                commonPassword: false,
                nistEntropyBits: 0,
                passwordLength: 0,
                shannonEntropyBits: 0,
                strengthCode: null,
                trigraphEntropyBits: null,
                charsets: null
            };

            // For undefined and if we get an empty value
            if (!currentPassword || !currentPassword.length) {
                if (this.trigraph) {
                    result.trigraphEntropyBits = 0;
                }

                return result;
            }

            result.commonPassword = this.checkCommonPasswords(currentPassword);
            result.charsets = this.charsetGroups(currentPassword);
            result.charsetSize = this.charsetSize(result.charsets);
            result.nistEntropyBits = this.nistScore(currentPassword);
            result.shannonEntropyBits = this.shannonScore(currentPassword);
            result.passwordLength = currentPassword.length;
            result.trigraphEntropyBits = this.checkTrigraph(currentPassword, result.charsetSize);
            result.strengthCode = this.determineStrength(result);

            return result;
        };

        /**
         * This checks to see if the password is in the common list of
         * passwords.  If in the common passwords lists, sets the warn flag.
         *
         * @param {string} passToCheck
         * @return {boolean}
         */
        this.checkCommonPasswords = function (passToCheck) {
            var commonPasswordsLength, commonPasswordsList, i;

            passToCheck = passToCheck.toLowerCase();

            if (this.commonPasswords && this.commonPasswords.length) {
                commonPasswordsList = this.commonPasswords;
                commonPasswordsLength = this.commonPasswords.length;

                for (i = 0; i < commonPasswordsLength; i += 1) {
                    if (commonPasswordsList[i] === passToCheck) {
                        return true;
                    }
                }

                return false;
            }

            return null;
        };


        /**
         * Compares the password to the trigraph to get the rating of how
         * guessable the password is.
         *
         * @param {string} passToCheck
         * @param {number} charsetSize
         * @return {number} entropy in bits
         */
        this.checkTrigraph = function (passToCheck, charsetSize) {
            var i, score, str;

            if (!this.trigraph) {
                return null;
            }

            score = 1;
            passToCheck = passToCheck.toLowerCase().replace(/[\W_]/gi, " ").trim();
            passToCheck = "_" + passToCheck + "_";

            for (i = 0; i < passToCheck.length - 2; i += 1) {
                str = passToCheck.substr(i, 3);

                if (this.trigraph[str]) {
                    // Less than fully random
                    score *= (1.0 - this.trigraph[str] / 10000) * charsetSize;
                } else {
                    // Fully random selection
                    score *= charsetSize;
                }
            }

            return Math.log2(score);
        };


        /**
         * Takes parameters and calculates the strength of a given password
         * from a list of codes set. This will use the trigraph entropy bits if
         * available and fallback to the shannon entropy bits which we should
         * always have.
         *
         * @param {Object} status
         * @return {string} strengthCode
         */
        this.determineStrength = function (status) {
            var entropyBits, strengthCode;

            strengthCode = "";

            if (status.trigraphEntropyBits) {
                entropyBits = status.trigraphEntropyBits;
            } else {
                entropyBits = status.shannonEntropyBits;
            }

            if (entropyBits <= 32) {
                // Keep out a typical attacker for minutes
                strengthCode = "VERY_WEAK";
            } else if (entropyBits <= 48) {
                // Crackable by a typical home computer in a week.
                strengthCode = "WEAK";
            } else if (entropyBits <= 64) {
                // A specialized computer could get this in one year.
                strengthCode = "REASONABLE";
            } else if (entropyBits <= 80) {
                // Resistant to a large, coordinated attack (botnet) for over a year.
                strengthCode = "STRONG";
            } else {
                // Nearly impossible to brute force, given more than all of the
                // computing power in the world, optimized algorithms,
                // specialized hardware and a thousand years.
                strengthCode = "VERY_STRONG";
            }

            return strengthCode;
        };


        /**
         * Calculates the NIST score based on the following rules from NIST
         * Special Publication 800-63 as of June 2004.
         *
         * * The entropy of the first character is four bits;
         * * The entropy of the next seven characters are two bits per
         *   character;
         * * The ninth through the twentieth character has 1.5 bits of
         *   entropy per character;
         * * Characters 21 and above have one bit of entropy per character.
         * * A "bonus" of six bits is added if both upper case letters and
         *   non-alphabetic characters are used.
         * * A "bonus" of six bits is added for passwords of length 1 through
         *   19 characters following an extensive dictionary check to ensure
         *   the password is not contained within a large dictionary.
         *   Passwords of 20 characters or more do not receive this bonus
         *   because it is assumed they are pass-phrases consisting of
         *   multiple dictionary words.
         *
         * According to Wikipedia, this score might not be overly useful.
         *
         * > Later research into human-selected password entropy using newly
         * > available real world data has demonstrated that the NIST scheme
         * > does not provide a valid metric for entropy estimation of
         * > human-selected passwords.
         *
         * @param {string} password
         * @return {number}
         */
        this.nistScore = function (password) {
            var len, score;

            len = password.length;
            score = 0;

            if (len > 20) {
                score += len - 20;
                len = 20;
            }

            if (len > 8) {
                score += 1.5 * (len - 8);
                len = 8;
            }

            if (len > 1) {
                score += 2 * (len - 1);
                len = 1;
            }

            if (len) {
                score += 4;
            }

            if (password.match(/[A-Z]/) && password.match(/[^A-Za-z]/)) {
                score += 6;
            }

            // Don't have a large dictionary, thus nobody gets the 6 bit
            // bonus.
            return score;
        };


        /**
         * Calls function to see if there are other characters used
         * which we don't explicitly check for.
         *
         * @param {string} passToCheck
         * @return {string} result
         */
        this.otherChars = function (passToCheck) {
            var chars, hash, pattern;

            chars = "";
            objectForEach(this.charsets, function (key, value) {
                chars += value;
            });
            pattern = safeCharMatcher(chars, "g");
            hash = {};
            passToCheck.replace(pattern, "").split("").forEach(function (letter) {
                hash[letter] = true;
            });

            return Object.keys(hash).join("");
        };


        /**
         * Calculate the Shannon Entropy value of a password.  This is really
         * the number of bits necessary to represent the possible characters
         * times the length of the password.
         *
         * Reworked from code under a MIT license.
         * entropy.js MIT License © 2014 James Abney http://github.com/jabney
         *
         * @param {string} pass Password to evaluate
         * @return {number} Bits of entropy
         */
        this.shannonScore = function (pass) {
            var freq, passLength, sum;

            /**
             * Counts how often each character appears in the passphrase.
             *
             * @return {Object} Treated like a hash, indexed by the character.
             */
            function countFrequencies() {
                var c, h, i;

                // Hash for collecting frequencies
                h = {};

                for (i = 0; i < passLength; i += 1) {
                    c = pass.charAt(i);

                    if (h[c]) {
                        h[c] += 1;
                    } else {
                        h[c] = 1;
                    }
                }

                return h;
            }

            sum = 0;
            passLength = pass.length;
            freq = countFrequencies();
            objectForEach(freq, function (key, value) {
                var score;

                score = value / passLength;
                sum -= score * Math.log2(score);
            });

            return sum * passLength;
        };
    }

    return PasswordStrength;
    // fid-umd post
}));
// fid-umd post-end
